File: 1-introdução/ProcessadorDePagamento/src/class/ProcessadorDePagamento.ts


export class ProcessadorDePagamento {
    // Método abstrato (simplificado): o usuário só precisa saber que "paga"
    pagar(valor: number, metodo: string) {
        if (metodo === "cartao") {
            this.processarPagamentoCartao(valor)
        } else if (metodo === "pix") {
            this.processarPagamentoPix(valor)
        }
    }

    private processarPagamentoCartao(valor: number) {
        console.log(`[CARTÃO] Validando... cobrando R$${valor}...`)
        this.mensagemConfirmacao()
    }

    private processarPagamentoPix(valor: number) {
        console.log(`[PIX] Validando... cobrando R$${valor}...`)
        this.mensagemConfirmacao()
    }

    mensagemConfirmacao() {
        setTimeout(() => {
            console.log(`transação confirmada. `)
        }, 1000)

    }

}

File: 3-POO-LOGIC-WITH-PYTHON/1-INTRODUCAO/CLASSE-PROCESSAR-PAGAMENTO/classes/Processador_de_pagamento.py


class Processador_de_pagamento:
# Método abstrato (simplificado): o usuário só precisa saber que "paga"
    def __init__(self):
        pass
    
    def pagar(self, valor:int, metodo_pagamento:str):
        # Validação básica
        if valor <= 0:
            raise ValueError("Valor deve ser maior que zero.")
        
        match metodo_pagamento:
            case "pix":
               self.__processar_pix(valor)
            case "cartao":
                self.__processar_cartao(valor)
            case _:
                raise ValueError("Método de pagamento inválido. Use 'pix' ou 'cartao'.")
            
    def __processar_pix(self, valor: float) -> dict:
        print(f"[PIX] Processando pagamento de R${valor:.2f}...")
        print("✅ PIX recebido com sucesso!")
        return {"sucesso": True, "metodo": "pix", "valor": valor}

    def __processar_cartao(self, valor: float) -> dict:
        print(f"[CARTÃO] Validando transação de R${valor:.2f}...")
        print("✅ Pagamento com cartão aprovado!")
        return {"sucesso": True, "metodo": "cartao", "valor": valor}